java/ql/src/Telemetry/ExternalAPI.qll

```diff
@@ -0,0 +1,52 @@
+/** Provides classes and predicates related to handling APIs from external libraries. */
+
+private import java
+private import APIUsage
+private import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * An external API from either the Java Standard Library or a 3rd party library.
+ */
+class ExternalAPI extends Callable {
+  ExternalAPI() { not this.fromSource() }
+
+  /** Holds true if this API is part of a common testing library or framework */
+  predicate isTestLibrary() { getDeclaringType() instanceof TestLibrary }
+
+  /**
+   * Gets information about the external API in the form expected by the CSV modeling framework.
+   */
+  string asCSV(ExternalAPI api) {
+    result =
+      api.getDeclaringType().getPackage() + ";?;" + api.getDeclaringType().getSourceDeclaration() +
+        ";" + api.getName() + ";" + paramsString(api)
+  }
+
+  /** Holds true if this API is not yet supported by existing CodeQL libraries */
+  predicate isSupported() { not supportKind(this) = "?" }
```

Comment:
I'd like to replace the entirety of `APIUsage.qll` with the following:
```suggestion
  /** Gets a node that is an input to a call to this API. */
  private DataFlow::Node getAnInput() {
    exists(Call call | call.getCallee().getSourceDeclaration() = this |
      result.asExpr().(Argument).getCall() = call or
      result.(ArgumentNode).getCall() = call
    )
  }
  
  /** Gets a node that is an output from a call to this API. */
  private DataFlow::Node getAnOutput() {
    exists(Call call | call.getCallee().getSourceDeclaration() = this |
      result.asExpr() = call or
      result.(PostUpdateNode).getPreUpdateNode().(ArgumentNode).getCall() = call
    )
  }
  
  /** Holds if this API has a supported summary. */
  predicate hasSummary() {
    this instanceof SummarizedCallable or
    TaintTracking::localAdditionalTaintStep(this.getAnInput(), _)
  }
  
  /** Holds if this API is a known source. */
  predicate isSource() { this.getAnOutput() instanceof RemoteFlowSource or sourceNode(this.getAnOutput(), _) }

  /** Holds if this API is a known sink. */
  predicate isSink() { sinkNode(this.getAnInput(), _) }

  /** Holds if this API is supported by existing CodeQL libraries, that is, it is either a recognized source or sink or has a flow summary. */
  predicate isSupported() { hasSummary() or isSource() or isSink() }
```

Reply:
Thanks for that!